### PR TITLE
Use system time for holiday menu screens

### LINKED ISSUE #36628: Use system time for holidays
I've noticed that we've been manually changing the title screens for the holiday variants. I don't think it's really necessary to make an entirely new patch update just for a small Easter egg. I have a better solution: system time. The game will pull from the system's internal clock, and will display whatever title screen based on that. I did a bit of digging to see if C++ supports pulling from system time, and, with my very little knowledge on programming code from scratch, here's what I got.

```
#include <ctime>

int main() {

time_t currentTime;
struct tm *local time;

time( &currentTime );
 localTime = localtime( &currentTime );

int month = localTime->tm_mon; 
init day = localTime->tm_mday;
init wday = localTime->tm_wday;


if((month = 0) && (day = 1))
	static const holiday current_holiday holiday::new_year;

if((month = 2) && (day ≥ 21) && (wday = 0))
	static const holiday current_holiday holiday::easter;

if((month = 3) && (day ≤ 7) && (wday = 0))
	static const holiday current_holiday holiday::easter;

if((month = 6) && (day = 4))
	static const holiday current_holiday holiday::independence_day;

if(month = 9)
	static const holiday current_holiday holiday::halloween;

if((month = 10) && (day = 25) && (wday = 4))
	static const holiday current_holiday holiday::thanksgiving;

if((month = 10) && (day = 26) && (wday = 4))
	static const holiday current_holiday holiday::thanksgiving;

if((month = 11) && (day ≤ 25))
	static const holiday current_holiday holiday::christmas;

ifElse
	static const holiday current_holiday holiday::none;
}
```

### PR DESCRIPTION
**Pull request by Drewscriver**
#### Summary
SUMMARY: Features "Fixes #36628, implementing use of system time for holidays"

#### Purpose of change
Fixes #36628, automatically setting holiday screens based on the local system date, instead of by manually updating the source code.  

Reduces code churn for holiday builds, and makes holiday easter egg screens viewable for users of future stable release versions.

#### Describe the solution
Change current_holiday from a static const to a member variable
Implement is_easter() as a member function
Implement get_holiday_from_time() as a member function, calling is_easter
 note that get_holiday_from_time() had to use preprocessor directives due to differing versions of localtime
Call get_holiday_from_time() when opening the screen

#### Describe alternatives you've considered
Letting someone else do it.

#### Testing
Set system date and checked that all holiday screens loaded  All did, except that in the process, I found that the ASCII art for independence day is messed up.

#### Additional context
std::localtime is non-intuitive, and Easter is now least favorite holiday.

### CODE REVIEW

**Comment by anothersimulacrum** [comment]
Put `fixes` in front of the issue it closes, or else github won't automatically close it when this PR is merged.

File: src/main_menu.cpp [hunk 1]
```diff
@@ -243,6 +242,52 @@ std::vector<std::string> main_menu::load_file( const std::string &path,
...
+    int e_b = year >> 2;
+    int e_c = ( e_b / 25 ) + 1;
+    int e_d = ( e_c * 3 ) >> 2;
+    int e_e = ( ( e_c << 3 ) + 5 ) / 25;
+    int e_f = ( e_a * 19 + e_d - e_e + 15 ) % 30;
+    int e_g = e_f + ( ( 29578 - e_a - ( e_f << 5 ) ) >> 10 );
+    int e_h = e_g - ( ( year + e_b - e_d + e_g + 2 ) % 7 );
+    int e_i = e_h >> 5; // is it April?
+    e_month = e_i + 3;
+    e_day = ( e_h & 31 ) + e_i;
```
Comment (anothersimulacrum) [inline src/main_menu.cpp:271]: Mind documenting what's going on here?

Comment (Drewscriver) [reply src/main_menu.cpp:271]: It's an implementation of Gauss's Easter Algorithm I found, but to be honest, I don't follow all the ins and outs of it myself; the explanations I've found are rather lengthy: https://math.stackexchange.com/questions/896954/decoding-gauss-easter-algorithm.  That's why I did make sure to test the Easter dates out to 2030.

Comment (anothersimulacrum) [reply src/main_menu.cpp:271]: Could you explain that in a comment, and possibly provide the same link?

Comment (Drewscriver) [reply src/main_menu.cpp:271]: Sure.  I wasn't sure we could link websites from source comments.

Comment (KorGgenT) [reply src/main_menu.cpp:271]: so to add on to this, if at all possible (and you know what they mean) can you make the variable names a little more verbose? and also make `const` the ones that do not change after initiation.

File: src/main_menu.cpp [hunk 2]
```diff
@@ -243,6 +242,55 @@ std::vector<std::string> main_menu::load_file( const std::string &path,
     return result;
 }
 
+holiday main_menu:: get_holiday_from_time()
+{
+    std::tm *local_time;
+
+    std::time_t current_time = std::time( nullptr );
+    local_time = std::localtime( &current_time );
```
Comment (anothersimulacrum) [inline src/main_menu.cpp:250]: Looks like this is needed to make lgtm.com happy.
```suggestion
    local_time = std::localtime_r( &current_time );
```

Comment (jbytheway) [inline src/main_menu.cpp:250]: In principle `localtime` can fail and return `nullptr`.  You might want to handle that.  Not sure if it really happens in practice.

**Comment by anothersimulacrum** [comment]
You need to remove the issue after the fixes too.
It should be something like `Fixes #12345`.

File: src/main_menu.cpp (hunk 2 above)
Comment (jbytheway) [reply src/main_menu.cpp]: I don't think `localtime_r` exists on Windows, so we probably have to use `localtime`.  I may be wrong, though.  Can't test it myself.

File: src/main_menu.cpp (hunk 1 above)
Comment (Drewscriver) [reply src/main_menu.cpp]: Did most of those; although I could not elaborate on the variable names, I used an implementation of the formula that is much more clearly comparable to the source algorithm.

File: src/main_menu.cpp (hunk 2 above)
Comment (Drewscriver) [reply src/main_menu.cpp]: Thanks, the code now checks localtime's return.

File: src/main_menu.cpp [hunk 3]
```diff
@@ -243,6 +242,77 @@ std::vector<std::string> main_menu::load_file( const std::string &path,
...
+        const int month = local_time.tm_mon + 1;
+        const int day = local_time.tm_mday;
+        const int wday = local_time.tm_wday;
+        const int year = local_time.tm_year + 1900;
+
+        /* check date against holidays */
+        if( month == 1 && day == 1 ) {
+            return holiday::new_year;
+        } else if( month == 3 ||
+                   month == 4 ) { // only run easter date calculation if currently March or April
```
Comment (anothersimulacrum) [inline src/main_menu.cpp:299]: Could you move this comment onto the line above or below? It's making this condition go across two lines and look really ugly.

File: src/main_menu.cpp [hunk 4]
```diff
@@ -243,6 +242,77 @@ std::vector<std::string> main_menu::load_file( const std::string &path,
...
+            return holiday::independence_day;
+        } else if( month == 10 && day >= 23 ) {
+            return holiday::halloween;
+        } else if( month == 11 && ( day >= 22 && day <= 28 ) && wday == 4 ) {
+            return holiday::thanksgiving;
+        } else if( month == 12 && day <= 25 ) {
+            return holiday::christmas;
+        }
+    }
+    return holiday::none; // fall through to here if localtime fails, or none of the day tests hit
```
Comment (anothersimulacrum) [inline src/main_menu.cpp:313]: I'd also like it if you could move this comment to the line above

File: src/main_menu.cpp [hunk 5]
```diff
@@ -243,6 +242,77 @@ std::vector<std::string> main_menu::load_file( const std::string &path,
     return result;
 }
 
+/* Calculate date of Easter - a direct implementation of the "Anonymous Gregorian" aka Meeus/Jones/Butcher Gregorian algorithm, given here:
+    https://en.wikipedia.org/wiki/Computus#Anonymous_Gregorian_algorithm */
```
Comment (anothersimulacrum) [inline src/main_menu.cpp:246]: I think this makes the comment prettier. If you don't agree, you can ignore this.
```suggestion
/*
 * Calculate date of Easter - a direct implementation of the "Anonymous Gregorian"
 * aka Meeus/Jones/Butcher Gregorian algorithm, given here:
 * https://en.wikipedia.org/wiki/Computus#Anonymous_Gregorian_algorithm
 */
```

**Comment by kevingranade** [comment]
So... there's a reason I said to have a table of dates you import, because I don't want anyone messing with gross calendar code.
Just grab a table of dates like this and embed it in the code: https://www.theholidayspot.com/easter/easter_dates.htm

File: src/main_menu.cpp [hunk 6]
```diff
@@ -243,6 +242,100 @@ std::vector<std::string> main_menu::load_file( const std::string &path,
...
+            case 2026:
+                return ( day == 5 );
+            case 2028:
+                return ( day == 16 );
+            case 2029:
+                return ( day == 1 );
+            case 2030:
+                return ( day == 21 );
+            default:
+                break;
```
Comment (anothersimulacrum) [inline src/main_menu.cpp:277]: Might look nicer
```suggestion
        switch( year ) {
            // *INDENT-OFF*
            case 2024: return ( day == 31 );
            case 2027: return ( day == 28 );
            default: break;
            // *INDENT-ON*
        }
        // month == 4 - april
    } else {
        // *INDENT-OFF*
        switch( year ) {
            case 2021: return ( day == 4 );
            case 2022: return ( day == 17 );
            case 2023: return ( day == 9 );
            case 2025: return ( day == 20 );
            case 2026: return ( day == 5 );
            case 2028: return ( day == 16 );
            case 2029: return ( day == 1 );
            case 2030: return ( day == 21 );
            default: break;
            // *INDENT-ON*
```

File: src/main_menu.cpp [hunk 7]
```diff
@@ -243,6 +242,100 @@ std::vector<std::string> main_menu::load_file( const std::string &path,
...
+
+        /* check date against holidays */
+        if( month == 1 && day == 1 ) {
+            return holiday::new_year;
+        }
+        // only run easter date calculation if currently March or April
+        else if( month == 3 || month == 4 ) {
+            if( is_easter( day, month, year ) ) {
+                return holiday::easter;
+            }
```
Comment (anothersimulacrum) [inline src/main_menu.cpp:320]:
```suggestion
        else if( month == 3 || month == 4 && is_easter( day, month, year ) ) {
            return holiday::easter;
```

File: src/main_menu.h [hunk 8]
```diff
@@ -5,11 +5,13 @@
 #include <cstddef>
 #include <string>
 #include <vector>
+#include <ctime>
```
Comment (anothersimulacrum) [inline src/main_menu.h:8]: Why do you have this include? You don't appear to need it here.

File: src/main_menu.cpp [hunk 9]
```diff
@@ -243,6 +242,90 @@ std::vector<std::string> main_menu::load_file( const std::string &path,
...
+        switch( year ) {
+            // *INDENT-OFF*
+            case 2021: return ( day == 4 );
+            case 2022: return ( day == 17 );
+            case 2023: return ( day == 9 );
+            case 2025: return ( day == 20 );
+            case 2026: return ( day == 5 );
+            case 2028: return ( day == 16 );
+            case 2029: return ( day == 1 );
+            case 2030: return ( day == 21 );
```
Comment (Night-Pryanik) [inline src/main_menu.cpp:267]: You could remove brackets here, they are redundant.

File: src/main_menu.h (hunk 8 above)
Comment (Drewscriver) [reply src/main_menu.h:8]: Looks leftover from a previous iteration.  Will remove.

File: src/main_menu.cpp [hunk 10]
```diff
@@ -243,6 +242,90 @@ std::vector<std::string> main_menu::load_file( const std::string &path,
...
+{
+    if( month == 3 ) {
+        switch( year ) {
+            // *INDENT-OFF*
+            case 2024: return day == 31;
+            case 2027: return day == 28;
+            default: break;
+            // *INDENT-ON*
+        }
+        // month == 4 - april
```
Comment (ZhilkinSerg) [inline src/main_menu.cpp:256]:
```suggestion
```

File: src/main_menu.cpp [hunk 11]
```diff
@@ -243,6 +242,90 @@ std::vector<std::string> main_menu::load_file( const std::string &path,
...
+    } else {
```
Comment (ZhilkinSerg) [inline src/main_menu.cpp:257]:
```suggestion
    } else if( month == 4 ) {
```

File: src/main_menu.cpp [hunk 12]
```diff
@@ -243,6 +242,89 @@ std::vector<std::string> main_menu::load_file( const std::string &path,
...
+            case 2025: return day == 20;
+            case 2026: return day == 5
+            case 2028: return day == 16;
+            case 2029: return day == 1;
+            case 2030: return day == 21;
+            default: break;
+            // *INDENT-ON*
+        }
+    }
+    // in practice, this should not be reached
```
Comment (ZhilkinSerg) [inline src/main_menu.cpp:271]:
```suggestion
```

File: src/main_menu.cpp [hunk 13]
```diff
@@ -243,6 +242,89 @@ std::vector<std::string> main_menu::load_file( const std::string &path,
...
+        const int day = local_time.tm_mday;
+        const int wday = local_time.tm_wday;
+        const int year = local_time.tm_year + 1900;
+
+        /* check date against holidays */
+        if( month == 1 && day == 1 ) {
+            return holiday::new_year;
+        }
+        // only run easter date calculation if currently March or April
+        else if( ( month == 3 || month == 4 ) && is_easter( day, month, year ) ) {
```
Comment (ZhilkinSerg) [inline src/main_menu.cpp:308]:
```suggestion
        else if( is_easter( day, month, year ) ) {
```

File: src/main_menu.cpp [hunk 14]
```diff
@@ -243,6 +242,89 @@ std::vector<std::string> main_menu::load_file( const std::string &path,
...
+            // *INDENT-ON*
+        }
+    } else if( month == 4 ) {
+        switch( year ) {
+            // *INDENT-OFF*
+            case 2021: return day == 4;
+            case 2022: return day == 17;
+            case 2023: return day == 9;
+            case 2025: return day == 20;
+            case 2026: return day == 5
```
Comment (ZhilkinSerg) [inline src/main_menu.cpp:263]:
```suggestion
            case 2026: return day == 5;
```

**Comment by nexusmrsep** [comment]
Fragment of this PR results in compilation errors under Code::Blocks project, as follows:

image

image

**Comment by ZhilkinSerg** [comment]
Which gcc version?

**Pull request merged**